src/services/access/useSessions.ts

```diff
@@ -0,0 +1,115 @@
+import { useQuery } from '@tanstack/react-query'
+import { z } from 'zod'
+
+import Api from 'shared/api'
+import { NetworkErrorObject } from 'shared/api/helpers'
+import { mapEdges } from 'shared/utils/graphql'
+
+const SessionSchema = z.object({
+  sessionid: z.number(),
+  ip: z.string().nullable(),
+  lastseen: z.string().nullable(),
+  useragent: z.string().nullable(),
+  type: z.string(),
+  name: z.string().nullable(),
+  lastFour: z.string(),
+})
+
+export type Session = z.infer<typeof SessionSchema>
+
+const UserTokenSchema = z.object({
+  id: z.string(),
+  name: z.string(),
+  type: z.string(),
+  lastFour: z.string(),
+})
+
+export type UserToken = z.infer<typeof UserTokenSchema>
+
+const RequestSchema = z.object({
+  me: z
+    .object({
+      sessions: z.object({
+        edges: z.array(
+          z.object({
+            node: SessionSchema,
+          })
+        ),
+      }),
+      tokens: z.object({
+        edges: z.array(
+          z.object({
+            node: UserTokenSchema,
+          })
+        ),
+      }),
+    })
+    .nullable(),
+})
+
+interface UseSessionsArgs {
+  provider: string
+}
+
+export function useSessions({ provider }: UseSessionsArgs) {
+  const query = `
+    query MySessions {
+      me {
+        sessions {
+          edges {
+            node{
+              sessionid
+              name
+              ip
+              lastseen
+              useragent
+              type
+              lastFour
+            }
+          }
+        }
+        tokens {
+          edges {
+            node {
+              type
+              name
+              lastFour
+              id
+            }
+          }
+        }
+      }
+    }
+  `
+
+  return useQuery({
+    queryKey: ['sessions', provider, query],
+    queryFn: ({ signal }) =>
+      Api.graphql({
+        provider,
+        query,
+        signal,
+      }).then((res) => {
+        const parsedRes = RequestSchema.safeParse(res?.data)
+
+        if (!parsedRes.success) {
+          return Promise.reject({
+            status: 404,
```

Comment:
wrote about this in one of Nick's reviews, but what are your thoughts on making this a 400 vs 404?

Reply:
I agree 404 seems off, but I think this would probably be a 5XX kind of problem most of the time it occurs. Maybe I'm wrong on that feeling though.